source/material_model/rheology/composite_viscous_creep.cc

```diff
@@ -0,0 +1,238 @@
+                                                             pressure,
+                                                             temperature,
+                                                             diffusion_creep_parameters,
+                                                             dislocation_creep_parameters,
+                                                             peierls_creep_parameters,
+                                                             min_viscosities[composition],
+                                                             max_viscosities[composition]);
+
+            strain_rate_residual = edot_and_deriv.first - edot_ii;
+            strain_rate_deriv = edot_and_deriv.second;
+
+            // If the strain rate derivative is zero, we catch it below.
+            if (strain_rate_deriv>std::numeric_limits<double>::min())
+              creep_stress -= strain_rate_residual/strain_rate_deriv;
+            stress_iteration += 1;
+          }
+
+        // The creep stress is not the total stress, so we still need to do a little work to obtain the effective viscosity.
+        // First, we compute the stress running through the strain rate limiter
+        const double lim_stress = 2.*min_viscosities[composition]*(edot_ii - creep_stress/(2.*max_viscosities[composition]));
+
+        // Now we return the viscosity using the total stress
+        return (creep_stress + lim_stress)/(2.*edot_ii);
+      }
+
+
+
+      template <int dim>
+      std::pair<double, double>
+      CompositeViscousCreep<dim>::compute_strain_rate_and_derivative (const double creep_stress,
+                                                                      const double pressure,
+                                                                      const double temperature,
+                                                                      const DiffusionCreepParameters diffusion_creep_parameters,
+                                                                      const DislocationCreepParameters dislocation_creep_parameters,
+                                                                      const PeierlsCreepParameters peierls_creep_parameters,
+                                                                      const double min_viscosity,
+                                                                      const double max_viscosity) const
+      {
+        const std::pair<double, double> diff_edot_and_deriv = diffusion_creep.compute_strain_rate_and_derivative(creep_stress, pressure, temperature, diffusion_creep_parameters);
+        const std::pair<double, double> disl_edot_and_deriv = dislocation_creep.compute_strain_rate_and_derivative(creep_stress, pressure, temperature, dislocation_creep_parameters);
+        const std::pair<double, double> prls_edot_and_deriv = peierls_creep.compute_strain_rate_and_derivative(creep_stress, pressure, temperature, peierls_creep_parameters);
```

Comment:
I think it would be really useful to output these three strain rates as an additional material model output. If we output the plastic and elastic strain rate as well, then we can effectively see what deformation mechanism is dominant at each point. Would it make sense to return these three values in this function, or write a separate function to do this?

Reply:
I've tentatively chosen to pass a strain-rate vector by reference to the compute_viscosity function, which gets filled with these strain rates and the strain rate of the max_limiter (the four strain rates add up to the total strain rate).